### PR TITLE
[BUGFIX][QGIS Server] GetLegendGraphic: if LAYERTITLE is false disable layer name in legend

### PR DESCRIPTION
**Pull request by rldhont**
If the layer has only one legend node, it is embedded in parent.
In QGIS Server the user can specify no layer title, so the layer title has not be displayed.

### CODE REVIEW

**Comment by rldhont** [comment]
Hi @mhugent and @elpaso,
This a customer request and I'd like some reviews.
Thanks,

**Review by elpaso — APPROVED** [review]
LGTM.
For the tests image comparisons, I think that there is already something implemented in other tests (not on the server), we should use the same image comparison algorithms for all tests. If not, you should consider moving the comparison method to the test utils library so that other tests can re-use it.

**Comment by luipir** [comment]
AFAIK the best person to ask about image compare utilities in core code/tests is @nyalldawson

**Comment by mhugent** [comment]
Looks good. +1 for merging once the tests are passing

**Comment by nyalldawson** [comment]
The render checker use looks correct. You'll need to generate a test mask when the travis build is fixed in order to get this to pass. To do this:

    scripts/parse_dash_results.py url_for_dash_test_result_page

**Review by sbrunner — APPROVED** [review]

File: tests/src/python/test_qgsserver.py [hunk 1]
```diff
@@ -431,6 +434,81 @@ def test_getLegendGraphics(self):
...
+            'SERVICE': 'WMS',
+            'VERSION': '1.3.0',
+            'REQUEST': 'GetLegendGraphic',
+            'FORMAT': 'image/png',
+            #'WIDTH': '20', # optional
+            #'HEIGHT': '20', # optional
+            'LAYER': u'testlayer+èé',
+            'LAYERTITLE': 'FALSE',
+        }
+        qs = '&'.join([u"%s=%s" % (k, v) for k, v in parms.iteritems()])
```
Comment (sbrunner) [inline tests/src/python/test_qgsserver.py]: Shouldn't be `itemes` instance of `iteritems`?

**Comment by rldhont** [comment]
@nyalldawson how to use scripts/parse_dash_results.py ? What is the dash file format ?

**Comment by nyalldawson** [comment]
@rldhont you just add the dash results url as an argument, eg

scripts/parse_dash_results.py "https://dash.orfeo-toolbox.org/testDetails.php?test=47533248&build=248193"

... except that that particular test is failing because of a python error, not a rendering difference.

**Comment by rldhont** [comment]
@nyalldawson thanks

**Comment by rldhont** [comment]
@nyalldawson I'd like to know, if you have an idea why I have the error
```
FileNotFoundError: [Errno 2] No such file or directory: '/tmp/WMS_GetLegendGraphic_test_result_diff.png'
```
I don't understand why the diff is not generated ?

File: tests/src/python/test_qgsserver.py [hunk 2]
```diff
@@ -431,6 +434,81 @@ def test_getLegendGraphics(self):
...
+            encoded_rendered_file = base64.b64encode(rendered_file.read())
+            message = "Image is wrong\n%s\nImage:\necho '%s' | base64 -d >%s/%s_result.png" % (
+                report, encoded_rendered_file.strip(), tempfile.gettempdir(), image
+            )
+
+        with open(os.path.join(tempfile.gettempdir(), image + "_result_diff.png"), "rb") as diff_file:
+            encoded_diff_file = base64.b64encode(diff_file.read())
+            message += "\nDiff:\necho '%s' | base64 -d > %s/%s_result_diff.png" % (
+                encoded_diff_file.strip(), tempfile.gettempdir(), image
+            )
```
Comment (sbrunner) [inline tests/src/python/test_qgsserver.py]: I think that you should try to change it on:
```
        diff_path = os.path.join(tempfile.gettempdir(), image + "_result_diff.png"
        if os.path.exists(diff_path):
            with open(diff_path), "rb") as diff_file:
                encoded_diff_file = base64.b64encode(diff_file.read())
                message += "\nDiff:\necho '%s' | base64 -d > %s/%s_result_diff.png" % (
                    encoded_diff_file.strip(), tempfile.gettempdir(), image
                )
        else:
            message += "Diff image '{}' not found.".format(diff_path)
```

**Comment by sbrunner** [comment]
You copy the methods `_result`, `_img_diff` and `_img_diff_error` from the `test_qgsserver_accesscontrol.py` file.
Shouldn't them bin in a new file like `qgsserver_helper.py` as function?

**Comment by rldhont** [comment]
@sbrunner I have copied theses methods to easily add tests. I have thought to create a `test_qgsserver_helper.py` in an other PR.

File: tests/src/python/test_qgsserver.py (hunk 2 above)
Comment (rldhont) [inline tests/src/python/test_qgsserver.py]: If I do so I just bypasses the test, isn't it ?

**Comment by sbrunner** [comment]
From the last Travis check you can get the generated an diff image by typing:
```
echo 'iVBORw0KGgoAAAANSUhEUgAAAC4AAAAjCAIAAABkRa++AAAACXBIWXMAAA3zAAAN8wEv/T3+AAABKElEQVRYhWP8//8/w+AATAPtAAQYdQo2MOoUbGDUKdjAqFOwgVGnYAOjTsEGRp2CDbCQoef58+enTp3i5uZ2cHBgYSHHBOo4ZfvOrfXtNX9Yvn95/53tP+fmddsVFRQHwCkbN28ob863z1eFcP/9+efoaXtg+xEFBQXKncJIfIPy////hhZ6xukiLOzMcMGXtz+y3JRZs3w95U4hIdnevn2bif0/sjsYGBiE5Xju3rtDuTtIc4qqquqFI9fQBD88//bz+296O4WRkTEhJe7K9mfIgrf3vsnJyqWKU0hIKxAQmRDy7N8tTlGmn19/f330P9I7qTC7ZGCcwsDAMHXq1ANH9jIzsYQFRwQFBVHFHWQ6hUZgEBX8o07BBkadgg2MOgUbGHUKNgAApPReTKXCbBsAAAAASUVORK5CYII=' | base64 -d > /tmp/WMS_GetLegendGraphic_test_layertitle_false_result.png

echo 'iVBORw0KGgoAAAANSUhEUgAAAC4AAAAjCAIAAABkRa++AAAACXBIWXMAAA9hAAAPYQGoP6dpAAAAkklEQVRYhe3XTQqAIBCG4YxgTuLZvFKdrZO0ayFY6LcZG3GC710K4YM/UWE/r8VH62zAEykoUlCkoEhBkYIiBUUKyhFl0z6QolQjh9HHl46SHe+5U5QUxUSj3qBqVqsl6aGMS0cxXIM2HSWfjGpkDiWvSpm+PcVfCh3/QYViu1/q94q5oPTbGzQ0UlCkoEhBOaLcf8YjI0EgpPgAAAAASUVORK5CYII='| base64 -d > /tmp/WMS_GetLegendGraphic_test_layertitle_false_result_diff.png  
```

**Comment by sbrunner** [comment]
=> 
wms_getlegendgraphic_test_layertitle_false_result

wms_getlegendgraphic_test_layertitle_false_result_diff

**Comment by rldhont** [comment]
Thanks, I was in getting and fixing it

**Comment by rldhont** [comment]
The QGIS Server tests has passed so I merge and backport. Thanks all

**Pull request merged**